File: archive/previous_campaign/tools/ghidra/scripts/SeedSymbols.java
// Ghidra headless script: seed known fn_XXXXXXXX symbols from a C source file.
// Invocation via analyzeHeadless:
//   -preScript SeedSymbols.java <path/to/gs_title.c>
// @category Pokemon

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.symbol.SourceType;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SeedSymbols extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 1) {
            println("usage: SeedSymbols <path/to/gs_title.c>");
            return;
        }
        String cPath = args[0];

        FunctionManager fnMgr = currentProgram.getFunctionManager();
        Pattern pat = Pattern.compile("\\bfn_([0-9A-Fa-f]{8})\\b");
        Set<String> seen = new HashSet<>();
        int created = 0;
        int renamed = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(cPath))) {
            String line;
            StringBuilder sb = new StringBuilder();
            while ((line = br.readLine()) != null) sb.append(line).append('\n');
            Matcher m = pat.matcher(sb);
            while (m.find()) {
                String hex = m.group(1).toLowerCase();
                if (!seen.add(hex)) continue;
                long addrVal = Long.parseLong(hex, 16);
                Address addr = currentProgram.getAddressFactory().getDefaultAddressSpace()
                        .getAddress(addrVal);
                if (currentProgram.getMemory().getBlock(addr) == null) continue;
                String name = "fn_" + hex.toUpperCase();
                Function fn = fnMgr.getFunctionAt(addr);
                if (fn == null) {
                    try {
                        fnMgr.createFunction(name, addr, null, SourceType.IMPORTED);
                        created++;
                    } catch (Exception e) {
                        println("create failed " + name + " @ " + addr + ": " + e);
                    }
                } else {
                    try {
                        fn.setName(name, SourceType.IMPORTED);
                        renamed++;
                    } catch (Exception e) {
                        println("rename failed " + name + " @ " + addr + ": " + e);
                    }
                }
            }
        }
        println("Seeded " + seen.size() + " addresses (created=" + created + ", renamed=" + renamed + ")");
    }
}

File: archive/previous_campaign/tools/ghidra/scripts/seed_symbols.py
# Ghidra headless script: seed known fn_XXXXXXXX symbols from src/game/gs_title.c
#
# Invocation (via analyzeHeadless):
#   -postScript seed_symbols.py <path/to/gs_title.c>
#
# Reads all `fn_XXXXXXXX` identifiers from the C file, extracts the 8-hex-digit
# address, and creates a named function at that address in the current program.
# If a function already exists, only the name is updated.
#@category Pokemon

import re
import sys

from ghidra.program.model.symbol import SourceType
from ghidra.program.model.address import AddressOutOfBoundsException


def seed(program, c_path):
    listing = program.getListing()
    fn_mgr = program.getFunctionManager()
    space = program.getAddressFactory().getDefaultAddressSpace()

    pattern = re.compile(r"\bfn_([0-9A-Fa-f]{8})\b")
    seen = set()

    with open(c_path, "r") as f:
        src = f.read()

    for m in pattern.finditer(src):
        hex_addr = m.group(1).lower()
        if hex_addr in seen:
            continue
        seen.add(hex_addr)

        name = "fn_" + hex_addr.upper()
        try:
            addr = space.getAddress(int(hex_addr, 16))
        except AddressOutOfBoundsException:
            continue

        block = program.getMemory().getBlock(addr)
        if block is None:
            continue

        fn = fn_mgr.getFunctionAt(addr)
        if fn is None:
            try:
                fn = fn_mgr.createFunction(name, addr, None, SourceType.IMPORTED)
            except Exception as e:
                print("create fail %s @ %s: %s" % (name, addr, e))
                continue
        else:
            try:
                fn.setName(name, SourceType.IMPORTED)
            except Exception as e:
                print("rename fail %s @ %s: %s" % (name, addr, e))

    print("Seeded %d unique fn_ addresses" % len(seen))


args = getScriptArgs()
if not args:
    print("usage: seed_symbols.py <path/to/gs_title.c>")
else:
    seed(currentProgram, args[0])
